[WIP] Refactor `EditorSceneFormatImporter`

atirutw opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

The `EditorSceneFormatImporter` is mostly used internally, but it is exposed, so I figured it should be properly documented and refactored for use from GDScript as it is very barely (and possibly poorly) documented.

It would be great if someone well-versed with the scene format import system would help me with this, because I am unfamiliar with the codebase and going by examples from classes that are already there.

Closes https://github.com/godotengine/godot-proposals/issues/7151

Tasks:
- [ ] Documentations
  - [x] Constants
  - [ ] Method descriptions (help wanted)
- [x] ~~Safeguards to prevent badly written plugins from crashing the editor~~ Seems to no longer be a problem
- [ ] Fix improperly exposed functions (help wanted)
- [ ] Investigate registering a format importer plugin not registering file types until a restart (separate issue/PR?)

atirutw committed:
Properly expose `ImportFlags` as enum

atirut-w commented:
Huh, does [this](https://github.com/godotengine/godot/actions/runs/6111177665/job/16585850230?pr=81421) have something to do with my commits or is it an existing problem?

AThousandShips commented:
It's unrelated, will just restart

aaronfranke commented on doc/classes/EditorSceneFormatImporter.xml:
```diff
@@ -44,17 +44,17 @@
 	<constants>
-		<constant name="IMPORT_SCENE" value="1">
+		<constant name="IMPORT_SCENE" value="1" enum="ImportFlags">
 		</constant>
-		<constant name="IMPORT_ANIMATION" value="2">
+		<constant name="IMPORT_ANIMATION" value="2" enum="ImportFlags">
```
In the near future, to solve https://github.com/godotengine/godot-proposals/issues/7494, I was thinking of refactoring these into an enum with many more options (scene, animation, node, mesh, material, etc). These would be a sequential enum (0, 1, 2, 3, 4) separate from the flags.

This doesn't technically block this change because we likely want to keep these values around anyway for compat, but making them a part of the enum seems like the wrong direction.

atirut-w replied:
They are defined as enums in the source code, though.

atirutw committed:
More descriptive error messages

atirutw committed:
Correct function signature for `import_scene()`

atirut-w commented:
Does someone know how to update the GDExtensions API?

atirutw committed:
Add description to `_get_extensions`

atirutw committed:
Properly expose `EditorSceneFormatImporter::get_import_options`

The implementation from `EditorImportPlugin` seems good enough, so I took the codes from that one.

It should ideally return an array of dictionaries, but I don't want to break more compat unless given a greenlight, so I just exposed the output argument for now.

atirut-w commented:
Oh.... so even exposing a new argument breaks compat...

atirut-w commented:
I think the GDE compat crisis is averted, but breaking GDS compat break is probably inevitable because `get_import_options` was originally exposed wrong.

atirutw commented:
Closing as stale and dead
